### PR TITLE
Finally create a mkdir utility

### PR DESCRIPTION
**Pull request by zachmullen**
ping @cpatrick

### CODE REVIEW

**Comment by brianhelba** [comment]
Have you seen https://bugs.python.org/issue25583 ? This may cause issues on Windows (if we care).

**Comment by zachmullen** [comment]
That problem doesn't bother me really, I don't think we ever try to create a directory that would be a drive name in Windows...

File: girder/utility/filesystem_assetstore_adapter.py [hunk 1]
```diff
@@ -59,12 +59,13 @@ def validateInfo(doc):
...
-        if not os.path.isdir(doc['root']):
-            try:
-                os.makedirs(doc['root'])
-            except OSError:
-                raise ValidationException(
-                    'Could not make directory "%s".' % doc['root'])
+
+        try:
+            mkdir(doc['root'])
+        except Exception:
```
Comment (cpatrick) [inline girder/utility/filesystem_assetstore_adapter.py]: Why are we `except`ing two different ways around this?

File: girder/utility/filesystem_assetstore_adapter.py [hunk 2]
```diff
@@ -85,14 +86,13 @@ def __init__(self, assetstore):
...
-            try:
-                os.makedirs(self.tempDir)
-            except OSError:
-                self.unavailable = True
-                logger.exception('Failed to create filesystem assetstore '
-                                 'directories %s' % self.tempDir)
-        elif not os.access(assetstore['root'], os.W_OK):
+        try:
+            mkdir(self.tempDir)
+        except OSError:
```
Comment (cpatrick) [inline girder/utility/filesystem_assetstore_adapter.py]: see above

File: girder/utility/plugin_utilities.py [hunk 3]
```diff
@@ -250,15 +250,13 @@ def getPluginDirs(curConfig=None):
...
-                os.makedirs(pluginDir)
-            except OSError:
-                if not os.path.exists(pluginDir):
-                    print(TerminalColor.warning(
-                        'Could not create plugin directory %s.' % pluginDir))
-
-                    failedPluginDirs.add(pluginDir)
+        try:
+            mkdir(pluginDir)
+        except Exception:
```
Comment (cpatrick) [inline girder/utility/plugin_utilities.py]: see above

File: girder/utility/filesystem_assetstore_adapter.py (hunk 1 above)
Comment (zachmullen) [inline girder/utility/filesystem_assetstore_adapter.py]: I think either one is fine, if we want them to be uniform which way would you prefer?

Comment (cpatrick) [inline girder/utility/filesystem_assetstore_adapter.py]: Ideally, we would `except` only the exceptions that can happen in those cases. I actually am not a big fan of `except Exception`.

Comment (zachmullen) [inline girder/utility/filesystem_assetstore_adapter.py]: That sounds good, I will make it `except OSError`.

**Comment by zachmullen** [comment]
@cpatrick I changed the `except`s.

File: girder/utility/__init__.py [hunk 4]
```diff
@@ -34,6 +37,30 @@ def camelcase(value):
...
+    """
+    Create a new directory or ensure a directory already exists.
+
+    :param path: The directory to create.
+    :type path: str
+    :param mode: The mode (permission mask) prior to applying umask.
+    :type mode: int
+    :param recurse: Whether intermediate missing dirs should be created.
+    :type recurse: bool
+    :param safe: Set to True to suppress the error if the dir already exists.
```
Comment (brianhelba) [inline girder/utility/__init__.py]: @zachmullen IMO, ``safe`` isn't a really clear name for a functionality which suppresses errors (I'd expect a safe mode to do more error checking).

[``os.makedirs`` in Python 3.2+ has the ``exist_ok`` parameter](https://docs.python.org/3/library/os.html#os.makedirs), which does the exact same thing as this parameter. How about just change the name to something like that?

**Comment by brianhelba** [comment]
LGTM other than the parameter name issues I mentioned inline.

File: girder/utility/__init__.py (hunk 4 above)
Comment (cpatrick) [inline girder/utility/__init__.py]: +1 to using exist_ok en lieu of safe.

**Pull request merged**